pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,278 @@
+	topologyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/topology/v1alpha1"
+)
+
+func (c *controller) reconcile(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	if ok, err := isSelectorWithoutConflict(partitionSet.Spec.ShardSelector, partitionSet.Spec.Dimensions); !ok {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionSetValid,
+			topologyv1alpha1.PartitionSetInvalidSelectorReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"",
+		)
+		// Deleting existing partitions when the new selector conflicts
+		if err = c.deletePartitions(ctx, partitionSet); err != nil {
+			return err
+		}
+		// No need to requeue if the selector conflicts
+		return nil
+	}
+	var labelSelector labels.Selector
+	if partitionSet.Spec.ShardSelector == nil {
+		labelSelector = labels.Everything()
+	} else {
+		var err error
+		labelSelector, err = metav1.LabelSelectorAsSelector(partitionSet.Spec.ShardSelector)
+		if err != nil {
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionSetValid,
+				topologyv1alpha1.PartitionSetInvalidSelectorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				err.Error(),
+			)
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionsReady,
+				topologyv1alpha1.ErrorGeneratingPartitionsReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"",
+			)
+			// Deleting existing partitions when the new selector is invalid
+			if err = c.deletePartitions(ctx, partitionSet); err != nil {
+				return err
+			}
+			// No need to requeue if the selector is not valid
+			return nil
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionSetValid)
+
+	shards, err := c.listShards(labelSelector)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
+		)
+		return err
+	}
+
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(partitionSet)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error generating PartitionSet key",
+		)
+		return err
+	}
+	oldPartitions, err := c.getPartitionsByPartitionSet(key)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"unable to retrieve associated Partitions",
+		)
+		return err
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
```

Comment:
it looks like we could get rid of this `if` statement since the `partition` function can already handle nil `shardSelectorLabels`

Reply:
I cannot reference `MatchLabels` on a nil pointer